src/main/java/com/atlassian/db/replica/internal/state/ConnectionState.java

```diff
@@ -210,7 +210,7 @@ private void closeConnection(
             } catch (Exception e) {
                 warnings.saveWarning(new SQLWarning(e));
             }
-            if (parameters.isReadOnly()) {
+            if (connection.isReadOnly()) {
```

Comment:
The `connection` <-> `parameters` dualism require sync. It'd be better if only one of them remained in existence.
I'll take a stab at it in the OOP refactor.

Reply:
Red `TestDualConnection.shouldResetReadOnlyModeBeforeReleasingConnection`:
```
Wanted but not invoked:
autoCommitAwareConnection.setReadOnly(false);
-> at com.atlassian.db.replica.api.TestDualConnection.shouldResetReadOnlyModeBeforeReleasingConnection(TestDualConnection.java:1435)

However, there were exactly 3 interactions with this mock:
autoCommitAwareConnection.getWarnings();
-> at com.atlassian.db.replica.internal.state.ConnectionState.closeConnection(ConnectionState.java:209)

autoCommitAwareConnection.isReadOnly();
-> at com.atlassian.db.replica.internal.state.ConnectionState.closeConnection(ConnectionState.java:213)

autoCommitAwareConnection.close();
-> at com.atlassian.db.replica.internal.state.ConnectionState.closeConnection(ConnectionState.java:217)
```

Hardening affected the mocking?

Reply:
Yes. The mock didn't have to have this state in sync :). So the test needs to improve with the code :).

Reply:
>  It'd be better if only one of them remained in existence.

🤞 
It will be awesome if it's possible. `parameters` keep more data than connection. It also knows if the property has been changed.